{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ExTdj_2SFlov",
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Migrate ligands\n",
    "\n",
    "> This notebook is intended to be run in Colab [![https img shields io badge colabs colab_notebooks 2Fcolab pyrosetta migrate_ligands ipynb f9ab00 logo googlecolab](https://img.shields.io/badge/colabs-colab_notebooks%2Fcolab--pyrosetta--migrate_ligands.ipynb-f9ab00?logo=googlecolab)](https://colab.research.google.com/github/matteoferla/pyrosetta_help/blob/master/colab_notebooks/colab-pyrosetta-migrate_ligands.ipynb)\n",
    "\n",
    "This notebook automates the migration of ligands from a homologue\n",
    "(provided or determined via blast) to the provided model, using PyRosetta. This includes:\n",
    "\n",
    "* parameterisation of unusual residues\n",
    "* poses are superimposed via neighbouring residues of the ones of ligands of interest (the equivalence of the neighbouring residues is determined by sequence aligment)\n",
    "* relaxation of the merge, with constraints based on hydrogen bonds form the donor pose\n",
    "\n",
    "This is for small molecules only. Not for NCAA or nucleic acids.\n",
    "\n",
    "The search for homologues does not use PyRosetta so can be stripped down.\n",
    "\n",
    "\n",
    "#### Resource location\n",
    "This notebook is from the repository [pyrosetta help](https://github.com/matteoferla/pyrosetta_help)\n",
    "<!--- silly badges: --->\n",
    "[![https img shields io github forks matteoferla pyrosetta_help label Fork style social logo github](https://img.shields.io/github/forks/matteoferla/pyrosetta_help?label=Fork&style=social&logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github stars matteoferla pyrosetta_help style social logo github](https://img.shields.io/github/stars/matteoferla/pyrosetta_help?style=social&logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github watchers matteoferla pyrosetta_help label Watch style social logo github](https://img.shields.io/github/watchers/matteoferla/pyrosetta_help?label=Watch&style=social&logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github last commit matteoferla pyrosetta_help logo github](https://img.shields.io/github/last-commit/matteoferla/pyrosetta_help?logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github license matteoferla pyrosetta_help logo github](https://img.shields.io/github/license/matteoferla/pyrosetta_help?logo=github)](https://github.com/matteoferla/pyrosetta_help/raw/master/LICENCE)\n",
    "[![https img shields io github release date matteoferla pyrosetta_help logo github](https://img.shields.io/github/release-date/matteoferla/pyrosetta_help?logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github commit activity m matteoferla pyrosetta_help logo github](https://img.shields.io/github/commit-activity/m/matteoferla/pyrosetta_help?logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github issues matteoferla pyrosetta_help logo github](https://img.shields.io/github/issues/matteoferla/pyrosetta_help?logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "[![https img shields io github issues closed matteoferla pyrosetta_help logo github](https://img.shields.io/github/issues-closed/matteoferla/pyrosetta_help?logo=github)](https://github.com/matteoferla/pyrosetta_help)\n",
    "\n",
    "#### Authors, funding and affiliations\n",
    "_Matteo Ferla_: Taylor group / Oxford Genomic Medicine theme, Wellcome Centre for Human Genetics, University of Oxford\n",
    "[WCHG](https://www.well.ox.ac.uk/people/matteo-ferla)\n",
    "[![https img shields io badge orcid 0000 0002 5508 4673 a6ce39 logo orcid](https://img.shields.io/badge/orcid-0000--0002--5508--4673-a6ce39?logo=orcid)](https://orcid.org/0000--0002--5508--4673) [![https img shields io badge google scholar gF bp_cAAAAJ success logo googlescholar](https://img.shields.io/badge/google--scholar-gF--bp_cAAAAJ-success?logo=googlescholar)](https://scholar.google.com/citations?user=gF--bp_cAAAAJ&hl=en) [![https img shields io twitter follow matteoferla label Follow logo twitter](https://img.shields.io/twitter/follow/matteoferla?label=Follow&logo=twitter)](https://twitter.com/matteoferla) [![https img shields io stackexchange stackoverflow r 4625475 logo stackoverflow](https://img.shields.io/stackexchange/stackoverflow/r/4625475?logo=stackoverflow)](https://stackoverflow.com/users/4625475) [![https img shields io stackexchange bioinformatics r 6322 logo stackexchange](https://img.shields.io/stackexchange/bioinformatics/r/6322?logo=stackexchange)](https://bioinformatics.stackexchange.com/users/6322) [![https img shields io badge email gmail informational logo googlemail](https://img.shields.io/badge/email-gmail-informational&logo=googlemail)](https://mailhide.io/e/Ey3RNO2G) [![https img shields io badge email Oxford informational logo googlemail](https://img.shields.io/badge/email-Oxford-informational&logo=googlemail)](https://mailhide.io/e/Y1dbgyyE)\n",
    "![Ox](https://upload.wikimedia.org/wikipedia/en/thumb/2/2f/University_of_Oxford.svg/132px-University_of_Oxford.svg.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "76GAr8tgbYab"
   },
   "outputs": [],
   "source": [
    "#@title Installation\n",
    "#@markdown Press the play button on the top right hand side of this cell\n",
    "#@markdown once you have checked the settings.\n",
    "#@markdown You will be notified that this notebook is not from Google, that is normal.\n",
    "!pip install rdkit-pypi rdkit-to-params biopython pyrosetta-help\n",
    "!pip install --upgrade plotly\n",
    "\n",
    "#@markdown Send error messages to errors.matteoferla.com for logging?\n",
    "#@markdown See [notebook-error-reporter repo for more](https://github.com/matteoferla/notebook-error-reporter)\n",
    "report_errors = False #@param {type:\"boolean\"}\n",
    "if report_errors:\n",
    "    !pip install notebook-error-reporter\n",
    "    from notebook_error_reporter import ErrorServer\n",
    "\n",
    "    es = ErrorServer(url='https://errors.matteoferla.com', notebook='fragmenstein')\n",
    "    es.enable()\n",
    "\n",
    "import os\n",
    "from importlib import reload\n",
    "import pyrosetta_help as ph\n",
    "from typing import *\n",
    "\n",
    "# Muppet-proofing: are we in colab?\n",
    "assert ph.get_shell_mode() == 'colab', 'This is a colab notebook, if running in Jupyter notebook, the installation is different'\n",
    "\n",
    "# ============================================================================\n",
    "#@markdown ### Use Google Drive\n",
    "#@markdown Optionally store your results in your google drive.\n",
    "#@markdown If `use_drive` is True, you will be prompted to give permission to use Google Drive\n",
    "#@markdown (you may be prompted to follow a link and possibly authenticate and then copy a code into a box)\n",
    "#@markdown —**_always_ remember to check strangers' code against data theft**:\n",
    "#@markdown e.g. search and look for all instances of `http`, `requests` and `post` in the code, and\n",
    "#@markdown make sure the creator is not typosquatting as someone else (e.g. username `Coogle`).\n",
    "use_drive = True  #@param {type:\"boolean\"}\n",
    "ph.mount_google_drive(use_drive)\n",
    "\n",
    "# ============================================================================\n",
    "#@markdown ### PyRosetta installation\n",
    "#@markdown This will install PyRosetta in this Colab notebook (~2–15 minutes depending on time of day),\n",
    "#@markdown but you will require a [PyRosetta licence](https://www.pyrosetta.org/home/licensing-pyrosetta)\n",
    "#@markdown (free for academics).\n",
    "#@markdown to speed things up _next_ time you can download a release into your Google Drive.\n",
    "#@markdown Use Google Drive for PyRosetta:\n",
    "\n",
    "#@markdown &#x1F44D; maybe faster next time\n",
    "\n",
    "#@markdown &#x1F44E; occupies some 10 GB, so you'll need to be on the 100 GB plan of Google Drive (it's one pound a month).\n",
    "\n",
    "download_to_drive = False #@param {type:\"boolean\"}\n",
    "download_path = '.' #@param {type:\"string\"}\n",
    "#@markdown If this is the next-time, `download_to_drive` and the credentials below will be ignored if\n",
    "#@markdown there's a release in `download_path`.\n",
    "\n",
    "#@markdown The following is not the real username and password. However, the format is similar.\n",
    "username = 'boltzmann'  #@param {type:\"string\"}\n",
    "username.strip().lower()\n",
    "password = 'constant'  #@param {type:\"string\"}\n",
    "#@markdown If yours are not the the academic credentials\n",
    "#@markdown disable this:\n",
    "hash_comparison_required = True #@param {type:\"boolean\"}\n",
    "#@markdown &#128544; THIS FLAG IS NOT PREVENTING YOU FROM USING PLAIN ROSETTA CREDENTIALS\n",
    "#@markdown AS THE CUSTOM ERROR SAYS! **REGULAR ROSETTA CREDENTIALS DO NOT WORK FOR PYROSETTA.**\n",
    "\n",
    "if download_to_drive and not use_drive:\n",
    "    raise ValueError('You said False to `use_drive` and True to `download_to_drive`? Very funny.')\n",
    "elif download_to_drive:\n",
    "    ph.download_pyrosetta(username=username,\n",
    "                          password=password,\n",
    "                          path=download_path,\n",
    "                          hash_comparison_required=hash_comparison_required)\n",
    "else:\n",
    "    pass\n",
    "\n",
    "ph.install_pyrosetta(username=username,\n",
    "                     password=password,\n",
    "                     path=download_path,\n",
    "                     hash_comparison_required=hash_comparison_required)\n",
    "reload(ph)\n",
    "\n",
    "# ??????? NOTE ??????????????????????????????????????????????????????????????????????\n",
    "# ? Note to code spies\n",
    "# ? this is a convoluted way to install pyrosetta via pyrosetta_help, due to options.\n",
    "# ? the quicker way is:\n",
    "# ?\n",
    "# ? >>> pip install pyrosetta-help\n",
    "# ? >>> install_pyrosetta -u xxx -p xxx\n",
    "# ??????????????????????????????????????????????????????????????????????????????????\n",
    "\n",
    "# disable as appropriate:\n",
    "!pip install py3Dmol\n",
    "# or:\n",
    "!pip install nglview\n",
    "from google.colab import output  # noqa (It's a colaboratory specific repo)\n",
    "output.enable_custom_widget_manager()\n",
    "\n",
    "# Colab still runs on 3.7\n",
    "# hack to enable the backport:\n",
    "import sys\n",
    "if sys.version_info.major != 3 or sys.version_info.minor < 8:\n",
    "    !pip install singledispatchmethod\n",
    "    import functools\n",
    "    from singledispatchmethod import singledispatchmethod  # noqa it's okay, PyCharm, I am not a technoluddite\n",
    "    functools.singledispatchmethod = singledispatchmethod\n",
    "    !pip install typing_extensions\n",
    "    import typing_extensions\n",
    "    import typing\n",
    "    for key, fun in typing_extensions.__dict__.items():\n",
    "      if key not in typing.__dict__:\n",
    "        setattr(typing, key, fun)\n",
    "\n",
    "# refresh imports\n",
    "import site\n",
    "site.main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "cUiIxUfRadq7"
   },
   "outputs": [],
   "source": [
    "#@title Start PyRosetta\n",
    "import pyrosetta\n",
    "import pyrosetta_help as ph\n",
    "\n",
    "#@markdown Do not optimise hydrogen on loading:\n",
    "no_optH = False #@param {type:\"boolean\"}\n",
    "#@markdown Ignore (True) or raise error (False) if novel residue (e.g. ligand) —  **don't tick this**.\n",
    "ignore_unrecognized_res=False  #@param {type:\"boolean\"}\n",
    "#@markdown Use autogenerated PDB residues are often weird (bad geometry, wrong match, protonated etc.): —best do it properly and parameterise it, so **don't tick this**.\n",
    "load_PDB_components=False  #@param {type:\"boolean\"}\n",
    "#@markdown Ignore all waters:\n",
    "ignore_waters=False  #@param {type:\"boolean\"}\n",
    "\n",
    "extra_options= ph.make_option_string(no_optH=no_optH,\n",
    "                                  ex1=None,\n",
    "                                  ex2=None,\n",
    "                                  mute='all',\n",
    "                                  ignore_unrecognized_res=ignore_unrecognized_res,\n",
    "                                  load_PDB_components=load_PDB_components,\n",
    "                                  ignore_waters=ignore_waters)\n",
    "\n",
    "\n",
    "# capture to log\n",
    "logger = ph.configure_logger()\n",
    "pyrosetta.init(extra_options=extra_options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "BlIL87pUbzsP"
   },
   "outputs": [],
   "source": [
    "#@title Upload acceptor model\n",
    "#@markdown Upload PDB file of the model into which the ligand(s) will be added.\n",
    "#@markdown this will be used in a first instance to extract the sequence for blasting.\n",
    "from google.colab import files\n",
    "#@markdown Alternatively, use the EBI-AlphaFold2 by providing a uniprot id\n",
    "#@markdown (leave blank for upload).\n",
    "uniprot= ''  #@param {type:\"string\"}\n",
    "\n",
    "#@markdown If the model has novel ligands, they will be loaded.\n",
    "#@markdown But to do this a residue type  (=topology) needs to be made or loaded.\n",
    "#@markdown These are saved as \"params files\".\n",
    "#@markdown These following options control both the \"acceptor\" and \"donor\" poses (if uploaded).\n",
    "\n",
    "#@markdown ## Params\n",
    "\n",
    "#@markdown * Some compounds are parameterised in the database folder of rosetta,\n",
    "#@markdown others in the PDB component database (if loaded).\n",
    "#@markdown * Uses the params defined in the cell of the acceptor pose.\n",
    "#@markdown * If there is no topology avalaible one will be made.\n",
    "#@markdown * If a params file is present in the working folder it will use it.\n",
    "#@markdown * See below or visit https://params.mutanalyst.com/ to generate them (upload the with the folder icon on the left).\n",
    "\n",
    "#@markdown This forces it (a bit silly):\n",
    "force_parameterisation = False  #@param {type:\"boolean\"}\n",
    "#@markdown If it needs to be parameterised make it protonated for pH 7?\n",
    "neutralize_params=True #@param {type:\"boolean\"}\n",
    "save_params=True #@param {type:\"boolean\"}\n",
    "\n",
    "\n",
    "#@markdown If a params file is present in the working folder it will use it.\n",
    "#@markdown Leave this blank... otherwise  (comma separated w/ no rando spaces):\n",
    "extra_params_files_to_use = '' #@param {type:\"string\"}\n",
    "extra_params = [f for f in extra_params_files_to_use.split(',') if f]\n",
    "use_all_folder_params= ''  #@param {type:\"boolean\"}\n",
    "if use_all_folder_params:\n",
    "  present_params = [filename for filename in os.listdir() if os.path.splitext(filename) == '.params']\n",
    "else:\n",
    "  present_params = []\n",
    "print('loading pose...')\n",
    "if uniprot:\n",
    "  acceptor_pose = ph.pose_from_alphafold2(uniprot)\n",
    "else:\n",
    "  uploaded = files.upload()\n",
    "  assert len(uploaded) ==1, 'wrong number of files (only one plz)'\n",
    "  pdbblock = list(uploaded.values())[0]\n",
    "  acceptor_pose = ph.ligands.load.parameterized_pose_from_pdbblock(pdbblock,\n",
    "                                     wanted_ligands = [],\n",
    "                                     force_parameterisation=force_parameterisation,\n",
    "                                     neutralize_params=neutralize_params,\n",
    "                                     save_params=save_params,\n",
    "                                     overriding_params=extra_params+present_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "6FS6gEm2xh31"
   },
   "outputs": [],
   "source": [
    "#@title Find related structures in the PDB\n",
    "#@markdown Chain of interest (always A in EBI-AF2):\n",
    "acceptor_chain = 'A' #@param {type:\"string\"}\n",
    "#@markdown Download a table of matches?\n",
    "download_csv = False #@param {type:\"boolean\"}\n",
    "\n",
    "print('blasting sequence...')\n",
    "chain_idx = ph.chain_letter_to_number(acceptor_chain, acceptor_pose)\n",
    "seq = acceptor_pose.chain_sequence(chain_idx)\n",
    "hunter = ph.LigandHunter(seq)\n",
    "print('Most common ligands:')\n",
    "for lig, c in hunter.get_most_common_ligands()[:10]:\n",
    "    print(lig, c, hunter.ligand_data[lig][0]['name'])\n",
    "df = hunter.to_dataframe()\n",
    "if download_csv:\n",
    "  df.to_csv('blast_hits.csv')\n",
    "  files.download('blast_hits.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "YqzkK5qzwxss"
   },
   "outputs": [],
   "source": [
    "#@title Choose the closest?\n",
    "\n",
    "#@markdown Find closest PDB with a given ligand (3-letter code).\n",
    "#@markdown ` CA` etc. are fine —The front space will be added.\n",
    "wanted_ligand = '' #@param {type:\"string\"}\n",
    "entry = hunter.get_pdb_entry_by_ligand(wanted_ligand)\n",
    "print(entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "qE9_sg7kvjUJ"
   },
   "outputs": [],
   "source": [
    "#@title load donor pose\n",
    "#@markdown If the result of the previous cell is good, leave these blank, \n",
    "#@markdown else provide a 4-letter PDB code and chain below:\n",
    "code = '' #@param {type:\"string\"}\n",
    "donor_chain = '' #@param {type:\"string\"}\n",
    "if not code:\n",
    "  code = entry['pdb_code']\n",
    "  donor_chain = entry['chain']\n",
    "\n",
    "#@markdown If the `cofactor_codes` from the output cell above are not okay,\n",
    "#@markdown declare which of its ligands you want (comma separated, \n",
    "#@markdown don't worry about spaces for ions, e.g. `GTP,MG`):\n",
    "wanted_ligands = '' #@param {type:\"string\"}\n",
    "if not wanted_ligands:\n",
    "  wanted_ligands = entry['cofactor_codes']\n",
    "else:\n",
    "  wanted_ligands = wanted_ligands.replace(' ', '').split(',')\n",
    "\n",
    "#@markdown If the wanted ligand is not quite the one available (e.g. GDP+Pi or GNP vs. GTP),\n",
    "#@markdown [Fragmenstein](https://github.com/matteoferla/Fragmenstein) is an option\n",
    "#@markdown ([example](https://blog.matteoferla.com/2020/07/switching-ligand-in-pdb-with.html))\n",
    "#@markdown —let me know and I'll add a cell.\n",
    "\n",
    "#@markdown Parameterisation options taken from previous cell.\n",
    "\n",
    "# recache\n",
    "if use_all_folder_params:\n",
    "  present_params = [filename for filename in os.listdir() if os.path.splitext(filename) == '.params']\n",
    "else:\n",
    "  present_params = []\n",
    "\n",
    "pdb_filename = ph.download_pdb(code)\n",
    "nicked = ph.LigandNicker(pdb_filename=pdb_filename,\n",
    "                         chain=donor_chain,\n",
    "                         neutralize_params=neutralize_params,\n",
    "                         wanted_ligands=wanted_ligands,\n",
    "                         force_parameterisation=force_parameterisation,\n",
    "                         overriding_params=extra_params+present_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "AlSA-NEdwJdi"
   },
   "outputs": [],
   "source": [
    "#@title Migrate the ligand\n",
    "\n",
    "#@markdown The steps done in this cell are several. Copypaste of docstring of migrate:\n",
    "#@markdown > This method aligns the sequences of the acceptor and donor pose.\n",
    "#@markdown > It finds the mapping of the neighbourhood of the wanted residues of the donor_pose\n",
    "#@markdown > It superimposes the poses by those residues.\n",
    "#@markdown > It adds the residues that need nicking.\n",
    "#@markdown > It adds constraints (optionally) based on the hydrogen bonding of the residues around the wanted residues.\n",
    "#@markdown > onto the ``.acceptor_pose``. \n",
    "#@markdown > It then optionally relaxes the neighbourhood.\n",
    "\n",
    "#@markdown (In the code the class is called LigandNicker, \n",
    "#@markdown which I realised afterwards `to nick` is British English slang\n",
    "#@markdown whereas most of the rest of code in American English as per custom)\n",
    "\n",
    "#@markdown leave the following as they are:\n",
    "constrained=True #@param {type:\"boolean\"}\n",
    "relaxed=True #@param {type:\"boolean\"}\n",
    "relax_radius=20  #@param {type:\"integer\"}\n",
    "relax_cycles=3  #@param {type:\"integer\"}\n",
    "nicked.migrate(acceptor_pose, acceptor_chain,\n",
    "               constrained=constrained,\n",
    "               relaxed=relaxed,\n",
    "               relax_radius=relax_radius,\n",
    "               relax_cycles=relax_cycles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "IPJfxCbz_AJx"
   },
   "outputs": [],
   "source": [
    "#@title Download PDB\n",
    "nicked.acceptor_pose.dump_pdb('added.pdb')\n",
    "files.download('added.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "TnVRgmbA_hI6"
   },
   "outputs": [],
   "source": [
    "#@title Upload to Michelanglo (optional)\n",
    "#@markdown [Michelanglo](https://michelanglo.sgc.ox.ac.uk/) is a website that\n",
    "#@markdown allows the creation, annotation and sharing of a webpage with an interactive protein viewport.\n",
    "#@markdown ([examples](https://michelanglo.sgc.ox.ac.uk/gallery)).\n",
    "#@markdown The created pages are private —they have a 1 in a quintillion change to be guessed within 5 tries.\n",
    "\n",
    "#@markdown Registered users (optional) can add interactive annotations to pages.\n",
    "#@markdown A page created by a guest is editable by registered users with the URL to it\n",
    "#@markdown (this can be altered in the page settings).\n",
    "#@markdown Leave blank for guest (it will not add an interactive description):\n",
    "\n",
    "username = ''  #@param {type:\"string\"}\n",
    "password = ''  #@param {type:\"string\"}\n",
    "\n",
    "import os\n",
    "assert not os.system(f'pip3 install michelanglo-api')\n",
    "import site\n",
    "site.main()\n",
    "from michelanglo_api import MikeAPI, Prolink\n",
    "if not username:\n",
    "  mike = MikeAPI.guest_login()\n",
    "else:\n",
    "  mike = MikeAPI(username, password)\n",
    "    \n",
    "page = mike.convert_pdb(pdbblock=ph.get_pdbstr(nicked.acceptor_pose),\n",
    "                        data_selection='ligand',\n",
    "                        data_focus='residue',\n",
    "                        )\n",
    "if username:\n",
    "     page.retrieve()\n",
    "     ligands_prolink = Prolink(text='migrated ligand(s)',\n",
    "                               focus='residue',\n",
    "                               selection=' or '.join(wanted_ligands))\n",
    "     \n",
    "     page.description = '## Description\\n\\n'\n",
    "     page.description += f'Model with {ligands_prolink} from {code} (chain {donor_chain}),'\n",
    "     page.description += 'namely:\\n\\n'\n",
    "     for lig in wanted_ligands:\n",
    "         page.description += '* '+ Prolink(text=lig,\n",
    "                                          focus='residue',\n",
    "                                          selection=lig) +'\\n'\n",
    "     page.commit()\n",
    "page.show_link()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "migrate_ligands.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}